### PR TITLE
Link extensions to options.

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: wire-compiler/src/main/java/com/squareup/wire/model/ProtoTypeName.java [hunk 1]
```diff
@@ -45,14 +45,24 @@
   public static final ProtoTypeName UINT32 = new ProtoTypeName("uint32");
   public static final ProtoTypeName UINT64 = new ProtoTypeName("uint64");
 
+  public static final ProtoTypeName FILE_OPTIONS = get("google.protobuf", "FileOptions");
```
Comment (JakeWharton) [inline wire-compiler/src/main/java/com/squareup/wire/model/ProtoTypeName.java]: Shouldn't this class be called `WireTypeName` to fit the naming scheme?

Comment (swankjesse) [inline wire-compiler/src/main/java/com/squareup/wire/model/ProtoTypeName.java]: Yes.

I started by trying to tease apart Java types like `com.squareup.cash.Payment` from Proto types like `squareup.cash.Payment`, and wanted to have a name to identify proto things vs. Java things.

As it turns out, that isn't really important yet, and `Wire` is proto.

Comment (swankjesse) [inline wire-compiler/src/main/java/com/squareup/wire/model/ProtoTypeName.java]: (will fix)

File: wire-compiler/src/test/java/com/squareup/wire/model/LinkerTest.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * limitations under the License.
+ */
+package com.squareup.wire.model;
+
+import java.util.LinkedHashSet;
+import java.util.Set;
+import org.junit.Test;
+
+import static java.util.Arrays.asList;
+import static org.junit.Assert.assertEquals;
```
Comment (JakeWharton) [inline wire-compiler/src/test/java/com/squareup/wire/model/LinkerTest.java]: Big PR to pull in AssertJ or Truth later?

Comment (JakeWharton) [inline wire-compiler/src/test/java/com/squareup/wire/model/LinkerTest.java]: I've gotten really good at IntelliJ's regex find-replace in path to do this for me...

Comment (swankjesse) [inline wire-compiler/src/test/java/com/squareup/wire/model/LinkerTest.java]: Hell yes. And Guava.

File: wire-compiler/src/main/java/com/squareup/wire/model/ProtoTypeName.java [hunk 3]
```diff
@@ -45,14 +45,24 @@
...
+  public static final ProtoTypeName MESSAGE_OPTIONS = get("google.protobuf", "MessageOptions");
+  public static final ProtoTypeName FIELD_OPTIONS = get("google.protobuf", "FieldOptions");
+  public static final ProtoTypeName ENUM_OPTIONS = get("google.protobuf", "EnumOptions");
+  public static final ProtoTypeName ENUM_VALUE_OPTIONS = get("google.protobuf", "EnumValueOptions");
+  public static final ProtoTypeName SERVICE_OPTIONS = get("google.protobuf", "ServiceOptions");
+  public static final ProtoTypeName METHOD_OPTIONS = get("google.protobuf", "MethodOptions");
```
Comment (JakeWharton) [inline wire-compiler/src/main/java/com/squareup/wire/model/ProtoTypeName.java]: Package scope? Or am I not foreseeing some use for these by plugins?

File: wire-compiler/src/test/java/com/squareup/wire/model/LinkerTest.java (hunk 2 above)
Comment (swankjesse) [inline wire-compiler/src/test/java/com/squareup/wire/model/LinkerTest.java]: (though I'd prefer to land https://github.com/square/wire/pull/222 first, which branches off this)

**Comment by JakeWharton** [comment]
LGTM :+1:

File: wire-compiler/src/main/java/com/squareup/wire/model/ProtoTypeName.java (hunk 3 above)
Comment (swankjesse) [inline wire-compiler/src/main/java/com/squareup/wire/model/ProtoTypeName.java]: Can do package scope. I started out thinking of this entire thing as an implementation detail; but I think it's going to be the thing that plugins are going to want to read.

**Pull request merged**